[RDY] Updated project plan with MoSCoW and other stuff

gfokkema opened the pull request.

gfokkema committed:
Updated project plan with MoSCoW and other stuff

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,79 @@ for programming because they facilitate exploratory programming and debugging.
+  * Indication of errors
+  * Integration with Eclipse
+* Should-have
+  * Syntax highlighting
```
I think this should be a must-have.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,79 @@ for programming because they facilitate exploratory programming and debugging.
+  * Save and load shell state
+* Could-have
+  * Integration with other IDE's (IntelliJ)
+    * Hover over variables to see value, type and others
```
Why is this a subbullet?

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,79 @@ for programming because they facilitate exploratory programming and debugging.
+track issues (all of this will be linked to Trello in one way or another).
+To facilitate this, we will use repositories from our GitHub organization
+[spoofax-shell](https://github.com/spoofax-shell).
+Each of us will then fork this repository to our private GitHub
```
`these repositories`

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,79 @@ for programming because they facilitate exploratory programming and debugging.
+of the development team. Once reviewed, the pull request can be changed
+to RDY, after which it can be merged.
+This process ensures that all code has been reviewed and tested before
+being merged into the final product, thereby always ensuring a working product.
```
I think we should also mention using continuous integration.

gfokkema replied:
good one

gfokkema replied:
fixed by now, github is outdated.

gfokkema committed:
Scrum update.

Hjdskes commented on project-plan.md:
```diff
@@ -82,3 +125,5 @@ Regular project deadlines:
 * 17-06-2016: SIG 2nd submission
 * 24-06-2016: Final presentation
+
+# Quality assurance
```
Perhaps this should be a subsection in the `Methodologies and tooling` section?

gfokkema replied:
I think you're right, deviating from the example report makes sense for this section.

Balletie committed:
pp.md: replace double space with tabs for nested lists

Not all markdown exporters understand double space, but all of them
understand tabs.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+* Must-have
+	* Interactive REPL
+		* Generated from language definition
+		* Should be language definition agnostic
```
Suggestions:
- Generated from language definition
- Implementation should be language-agnostic (e.g. work for any language defined in Spoofax)

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+as specified following the MoSCoW method:
+
+* Must-have
+	* Interactive REPL
```
The whole goal is to make an interactive REPL, so I think we can remove this bullet and lift all the subbullets a level higher.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+* Should-have
+	* Save and load shell state
+* Could-have
+	* Integration with other IDE's (IntelliJ)
```
`IDEs`

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+from the backlog will be assigned to project members, based on their
+prioritization and their feasibility for this sprint.
+Each sprint will end with a meeting in which we discuss to what degree
+we have achieved our sprint planning and how we can improve on that.
```
A summary of these meetings will be discussed with Hendrik on the following Monday during our weekly meetings?

I think we need to consider how we can combine these sessions with the weekly meetings.

gfokkema replied:
We need to come back to this issue.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+today, as well as any issues that could be detrimental to the weekly sprint goal.
+
+We will use a pull-based development model using GitHub in conjunction with
+the SCRUM methodology. GitHub Issues will be used to
```
You can remove this Scrum information here; it's already mentioned above.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
 (e.g. this document, the research project, et cetera) supporting our project.
 
-Since Spoofax is an already existing project, we will use whichever tools are
+A pull request will be designated a status tag WIP, RFC or RDY:
```
Can we make these tags bold?

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
-Since Spoofax is an already existing project, we will use whichever tools are
+A pull request will be designated a status tag WIP, RFC or RDY:
+
+* WIP: Work in progress
```
Suggestion:
- WIP - Work In Progress: a PR tagged WIP is there to gather feedback; it will still change so it should not be reviewed in-depth.
- RFC - Ready for Comments: a PR tagged RFC is done as far as the implementation goes: the functionality is there, it is tested, documented and only a (thorough) review is left. If there are too many comments that cannot be resolved within reasonable time, the PR should be demoted to WIP.
- RDY - Ready: a PR tagged RDY has been reviewed by at least one other reviewer, who deems it ready for merge. It is left open a short while to give the remaining team member time to review, after which it is merged.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+and that the pull request passes continuous integration testing.
+Pull requests with status RFC should be reviewed by at least one other member
+of the development team. Once reviewed, the pull request can be changed
+to RDY, after which it can be merged.
```
The above section can be removed if you apply ^.

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
+to RDY, after which it can be merged.
+This process ensures that all code has been reviewed and tested before
+being merged into the final product, thereby always ensuring a working product.
+
```
I think we should add a section here containing the following points (kindly taken from [Neovim](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md)):

## Submitting
- Make it clear in the issue tracker what you are working on.
- Be descriptive in your pull request description: what is it for, why is it needed, etc **or** link to the user story.
- Do not make cosmetic changes to unrelated files or sections in the same pull request. This creates noise, making reviews harder to do.

## Branching & history
- Do not work on your PR on the master branch, use a feature branch instead.
- Rebase your feature branch onto (upstream) master before opening the PR.
- Keep up to date with changes in (upstream) master so your PR is easy to merge.
- Try to actively tidy your history: combine related commits with interactive rebasing, separate monolithic commits, etc. If your PR is still [WIP], feel free to force-push to your feature branch to tidy your history.

Perhaps we should also have a section about [commit guidelines](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#commit-guidelines).

Hjdskes commented on project-plan.md:
```diff
@@ -24,50 +24,95 @@ for programming because they facilitate exploratory programming and debugging.
 * Maven for the build environment;
-* JUnit for unit tests.
+* JUnit for unit tests;
+* TravisCI for continous integration
```
`continuous`

Unia commented:
For the features, you can also take a look at my section `What is a REPL?`.

gfokkema committed:
Rewrite MoSCoW, fix Jente's comments.

Unia commented:
The quality assurance section still needs to be written. When that is done, I think this can be  merged.

Hjdskes committed:
Project plan: provide a section on quality assurance

Unia commented:
I have added a section on quality insurance. @gfokkema @Balletie please review.

Balletie commented on project-plan.md:
```diff
@@ -109,9 +110,37 @@ already being used. This means that we will use:
-# Quality assurance
+## Quality assurance
+
+There are two types of qualit to be assured:
```
quality

Balletie commented on project-plan.md:
```diff
@@ -109,9 +110,37 @@ already being used. This means that we will use:
+
+Further towards the end of the development period, we will ask one or two
+members of the client's research group to develop a (toy) language in Spoofax in
+combination with our REPL. This is to gather feedback from someone who has not
```
This is too much to ask from someone. I think you should rather ask them to use the REPL for one of the _existing_ (smaller/toy) languages, and maybe perform some changes to the language to see how the REPL helps in performing that change.

Hjdskes replied:
Good catch, that's actually what I was trying to say :P
